fdbserver/VersionedBTree.actor.cpp

```diff
@@ -7064,7 +7066,12 @@ class KeyValueStoreRedwoodUnversioned : public IKeyValueStore {
 		}
 
 		if (rowLimit > 0) {
-			wait(cur.seekGTE(keys.begin));
+			f = cur.seekGTE(keys.begin);
+			if (!isLocked && !f.isReady()) {
```

Comment:
Is checking for the future being ready enough to check if it's the cached case? Is it possible to get the future ready immediately in non-cached case?

Reply:
From my understanding, if the seek function completes immediately then it indicates there is no retrieving data from the disk. I think if it is able to complete immediately then maybe it's also fine even if it needs to visit the disk?

Reply:
Unless my understanding of flow futures is wrong, if the future is ready then whatever operation the future was doing either completed or got an error.

Reply:
This brings up a good point in that you should still wait on the future even if it is ready and you don't acquire the lock, to ensure any errors the read got are propagated properly.

Reply:
Yes, it's correct that the future will be ready when it has completed and we should wait on it either ways I believe.

Reply:
I am not 100% sure of the locking logic here, but was just thinking if invoking the seek itself without a lock could break any semantics?

Reply:
Sounds good. I will make that change.

Reply:
Well actually. I think neelam's right, because the future returned isn't the cache future, it's the cursor future. So that actor couldn't start and complete between the call to instantiate the actor, and the next line checking if the future is already ready. So I think you'll never see the case where f is ready.
Because it doesn't change the behavior, it won't break any semantics, but it also will never actually skip the lock acquire.
To do it properly, you'd basically have to pass something down the call stack from here to the point where it actually checks the cache future in IPager::readPage, and at the point where you have a read outstanding, you'd have to check if it's ready or not, and if it isn't already ready, acquire the lock there?